### PR TITLE
[ENH] Apply forecasters depending on series type (e.g., intermittent) compositor

### LINKED ISSUE #6279: [ENH] Composition forecaster that clusters time series and forecasts by cluster
**Is your feature request related to a problem? Please describe.**
Currently, there is no feature which estimates (provides) indicators based on past behavior or trends of a time series data set. Visual inspection is required before deciding how to best choose the appropriate preprocessing and model in order to produce forecast, be in direct or recursive. 

**Describe the solution you'd like**
To have a feature which estimate the forecastability of a time series by labeling them as smooth, erratic, lumpy, intermediate, unclassified (or just integer based labels. Same as what the clusterers of `sktime` are providing now via `get_fitted_params`). There are already existing studies that shows can this can be implemented for demand forecasting. But this may lead to misclassification if data contains both positive and negative values (see calculation from links below). Maybe passing the data through a `MinMaxScaler` solves the issue, any thoughts?

- [Measurement of ADI and CV²](https://frepple.com/blog/demand-classification/)

Another website discussing measurement for forecastability:
- [Approximate Entropy and Sample Entropy](https://ming-zhao.github.io/Business-Analytics/html/docs/time_series/stationarity.html)

Finally base on the result this feature provides, the user decides whether and how to preprocess, aggregate multiple time series to increase forecastability, or model the data, etc.

### PR DESCRIPTION
**Pull request by shlok191**
#### What does this implement/fix? Explain your changes.
This implementation introduces the `TransformSelectForecaster` which accepts a series-to-primitives transformer and a dictionary of forecasters as initialization parameters. This feature will allow users to categorize their input data through a transformer and then apply specific forecasters best suited for the classification categories.

#### What should a reviewer concentrate their feedback on?

A reviewer should most focus on the `_category_compositor.py` file in the `sktime/forecasting/compose` directory! It might also be helpful to take a look at the `ADICVTransformer` as it is a default parameter for this compositor.

#### Did you add any tests for the change?

I have not implemented any tests yet, but I am planning on adding them once I can get more feedback about the implementation and some advice on good test cases for this particular function!

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by shlok191** [comment]
@fkiraly, Hello Franz, I've made the changes that we discussed prior! I have made the series-to-primitives transformer a passable parameter and attempted to generalize the model. I had two questions I wanted to discuss though:

- For the passed transformer, is it safe for me to assume that the first column from a `transform` function will always include the data I require to choose a forecaster?

- For the test cases, would you have an idea about what might be good scenarios? I can think about instances that might lead to clear failures (like not passing in all required forecasters) but I think there can be better ones! 

Thanks a lot for the help!

File: sktime/forecasting/compose/_category_compositor.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+        X : Pd.Series | None, optional (default=None)
+            No exogenous variables are used for this.
+
+        Returns
+        -------
+        self : reference to self
+        """
+        # reset the state of the forecaster
+        self.reset()
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:92]: input checks and logic does not need to be carried out here. This is done in the boilerplate layer, in `BaseForecatser.fit`.

File: sktime/forecasting/compose/_category_compositor.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+            Exogeneous time series for the forecast
+
+        Returns
+        -------
+        y_pred : sktime time series object
+            should be of the same type as seen in _fit, as in "y_inner_mtype" tag
+            Point predictions
+        """
+        # Check if we have a fited forecaster to the data
+        self.check_is_fitted()
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:141]: same

File: sktime/forecasting/compose/_category_compositor.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+        # passing time series through the provided transformer!
+        self.category_ = self.transformer.fit_transform(y=y).iloc[0, 0]
+
+        # check if we have an available forecaster
+        assert (
+            self.category_ in self.forecasters
+        ), f"Forecaster not provided for given time series of type {self.category_}."
+
+        self.chosen_forecaster_ = self.forecasters[self.category_]
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:105]: makes sense, I wonder whether we can make it more robust by allowing the user to specify a `fallback_forecaster`, e.g., if the category is sth else. I would make it the third, optional argument, and use it as a default in a dict.get if it is not `None` - otherwise, raise a `ValueError` if no fallback is provided and the category is not found.

File: sktime/forecasting/compose/_category_compositor.py (hunk 1 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:92]: Ah, got it. Removing the checks from all relevant places!

File: sktime/forecasting/compose/_category_compositor.py (hunk 3 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:105]: I think that's definitely a good idea, I'll add it to the code :)

File: sktime/forecasting/compose/_category_compositor.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+        Implements default behaviour of calling update and predict sequentially, but can
+        be overwritten by subclasses to implement more efficient updating algorithms
+        when available.
+        """
+        self.update(y, X, update_params=update_params)
+        return self.predict(fh, X)
+
+    @classmethod
+    def get_test_params(cls, parameter_set="default"):
```
[the remark this answers is not in the document]
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:199]: Adding in test cases!

File: sktime/forecasting/compose/_category_compositor.py (hunk 1 above)
Comment (fkiraly) [reply sktime/forecasting/compose/_category_compositor.py:92]: have you done that? still seems like they are there

File: sktime/forecasting/compose/_category_compositor.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        self : reference to self
+        """
+        # reset the state of the forecaster
+        self.reset()
+
+        # asserting validity of input
+        assert y is not None
+
+        # passing time series through the provided transformer!
+        self.category_ = self.transformer.fit_transform(y=y).iloc[0, 0]
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:98]: Should this not be `X=y`?

**Review by fkiraly — CHANGES_REQUESTED** [review]
You need to make sure the estimator is exported in a public module, or tests will not be run. E.g., import to `compose.__init__.py`.

This should also highlight other issues - you can do the same by running `check_estimator`.

File: sktime/forecasting/compose/_category_compositor.py (hunk 5 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py]: Ah, made this fix in the new commit!

File: sktime/forecasting/compose/_category_compositor.py (hunk 1 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py]: I was at work at the time, but I've made the changes now

**Comment by shlok191** [comment]
@fkiraly, I am running the pytests locally and did not encounter the issues I'm seeing online. Could you please tell me if I'm doing something wrong with the soft dependencies? Thanks!

File: sktime/forecasting/compose/tests/test_category_compositor.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+
+@pytest.mark.parametrize(
+    "forecasters, series_generator, chosen_forecaster",
+    [
+        (
+            {
+                "smooth": NaiveForecaster(),
+                "intermittent": SARIMAX(),
```
Comment (fkiraly) [inline sktime/forecasting/compose/tests/test_category_compositor.py:22]: this requires a soft dependency, how about you use `Croston` instead, which is better for intermittent demand anyway?

File: sktime/forecasting/compose/tests/test_category_compositor.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    assert type(compositor.chosen_forecaster_) is type(chosen_forecaster)
+
+
+@pytest.mark.parametrize(
+    "forecasters, series_generator, fallback_forecaster",
+    [
+        (
+            {
+                "smooth": NaiveForecaster(),
+                "intermittent": SARIMAX(),
```
Comment (fkiraly) [inline sktime/forecasting/compose/tests/test_category_compositor.py:64]: same

File: sktime/forecasting/compose/_category_compositor.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+            Parameters to create testing instances of the class
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            `MyClass(**params)` or `MyClass(**params[i])` creates a valid test instance.
+            `create_test_instance` uses the first (or only) dictionary in `params`
+        """
+        param1 = {
+            "forecasters": {
+                "smooth": NaiveForecaster(),
+                "erratic": PolynomialTrendForecaster(),
+                "intermittent": SARIMAX(),
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:237]: this requires a soft dependency, how about you use `Croston` instead, which is better for intermittent demand anyway?

File: sktime/forecasting/compose/_category_compositor.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+from sktime.transformations.base import BaseTransformer
+from sktime.transformations.series.adi_cv import ADICVTransformer
+
+adi_cv_transformer = ADICVTransformer(features=["class"])
+
+
+class CategoryCompositor(BaseForecaster):
+    """Compositor that utilizes varying forecasters by time series data's nature.
+
+    Applies a series-to-primitives transformer on a given time series and utilizes
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:19]: can you be more precise regarding what this does?

File: sktime/forecasting/compose/_category_compositor.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        which can be used to quantify a choice of forecaster for the time series.
+
+    fallback_forecaster : sktime forecaster | None, Optional
+        A fallback forecaster that will be used if the category generated by
+        the transformer does not match any of the given forecasters.
+
+    Raises
+    ------
+    AssertionError: If a valid transformer (an instance of BaseTransformer)
+    is not passed or if valid forecasters (instances of BaseForecaster) is not given.
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:40]: an example would be nice - you could simply code from one of the test cases

File: sktime/forecasting/compose/_category_compositor.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    """
+
+    _tags = {
+        "y_inner_mtype": "pd.Series",
+        "X_inner_mtype": "pd.Series",
+        "scitype:y": "univariate",
+        "ignores-exogeneous-X": True,
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:47]: tags should be set differently - I would set scitype:y to both, ignores X to False; the inner mtypes to `pd.DataFrame` to handle multivariate data.

the capabilities, including "handles missing data" should be set so a capability is true if and only if all forecasters have the capability.

There is another thing we should do once this is sorted and tests pass, to handle global forecasters and hierarchical series - let´s make a note for it now, and I will explain it later once the simple case works.

**Comment by fkiraly** [comment]
> I am running the pytests locally and did not encounter the issues I'm seeing online. Could you please tell me if I'm doing something wrong with the soft dependencies? Thanks!

Which failures precisely, can you describe?

**Comment by fkiraly** [comment]
Re failures currently, I would suggest you use `check_estimator` to debug them, see guide here:
https://www.sktime.net/en/latest/developer_guide/add_estimators.html#using-the-check-estimator-utility

Feel free to ask in the `dev-chat` discord channel if you get stuck

**Comment by shlok191** [comment]
@benHeid, @astrogilda Hey guys! I'm nearly done working on this PR.

I've received some great feedback from Franz which I've tried to incorporate, and I'll really appreciate it if y'all could take a look at my code as well. 

Thank you so much!

**Comment by fkiraly** [comment]
Sure! short remark, the "protocol" for pull request review is that you should comment on the open discussions, one of:

* explain how addressed
* argue that needs not be addressed
* request information or discussion

Until consensus is reached.

File: sktime/forecasting/compose/_category_compositor.py [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+from sktime.forecasting.croston import Croston
+from sktime.forecasting.naive import NaiveForecaster
+from sktime.forecasting.trend import PolynomialTrendForecaster
+from sktime.transformations.base import BaseTransformer
+from sktime.transformations.series.adi_cv import ADICVTransformer
+
+adi_cv_transformer = ADICVTransformer(features=["class"])
+
+
+class CategoryCompositor(BaseForecaster):
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:18]: I think the name is a bit non-self-explanatory.

How about `ForecastByGroup`, or `ByGroupForecaster`?

File: sktime/forecasting/compose/_category_compositor.py [hunk 13]
```diff
@@ -0,0 +1,479 @@
...
+        "enforce_index_type": None,
+        "authors": ["shlok191"],
+        "maintainers": ["shlok191"],
+        "python_version": None,
+    }
+
+    def __init__(
+        self,
+        forecasters,
+        transformer=adi_cv_transformer,
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:86]: mutable objects should never be used as defaults in python methods or classes!

The reason is, if the object is mutated, then the default for any future calls is also mutated.

The pythonic way to do is is to have the defalut = `None`, and then check if is `None`, then set the param to the default.

Now, in `sklearn` API, params should be written to self and never overwritten, so you ought to write the default or transformer to `self.transformer`.

File: sktime/forecasting/compose/_category_compositor.py [hunk 14]
```diff
@@ -0,0 +1,479 @@
...
+        fallback_forecaster=None,
+    ):
+        # saving arguments to object storage
+        self.transformer = transformer
+        self.forecasters = forecasters
+        self.fallback_forecaster = fallback_forecaster
+
+        self.transformer_ = copy.deepcopy(transformer)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:94]: don't copy, but do `transformer.clone()`.

File: sktime/forecasting/compose/_category_compositor.py [hunk 15]
```diff
@@ -0,0 +1,479 @@
...
+        # All checks OK!
+
+        # Assigning all capabilities on the basis of the capabilities
+        # of the passed forecasters
+        true_if_all_tags = {
+            "ignores-exogeneous-X": False,
+            "X-y-must-have-same-index": False,
+            "enforce_index_type": False,
+            "handles-missing-data": False,
+            "capability:insample": False,
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:113]: these should be set to `True`, to signifiy that it might be, depending on the components used.

File: sktime/forecasting/compose/_category_compositor.py [hunk 16]
```diff
@@ -0,0 +1,479 @@
...
+            "capability:pred_int": False,
+            "capability:pred_int:insample": False,
+        }
+
+        # Extrapolating values for flags that should be True if they are
+        # True for all of the given forecasters!
+        for tag in true_if_all_tags.keys():
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:120]: this block of code is very long. Why not use `self._anytagis_then_set`?

File: sktime/forecasting/compose/_category_compositor.py [hunk 17]
```diff
@@ -0,0 +1,479 @@
...
+            if self.fallback_forecaster is None:
+                raise ValueError(
+                    "Forecaster not provided for given"
+                    + f"time series of type {self.category_}"
+                    + "and no fallback forecaster provided to use for this case."
+                )
+
+            # Adopt the fallback forecaster if possible
+            else:
+                self.chosen_forecaster_ = copy.deepcopy(self.fallback_forecaster)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:237]: `clone`, not deepcopy

File: sktime/forecasting/compose/_category_compositor.py [hunk 18]
```diff
@@ -0,0 +1,479 @@
...
+        Example:
+
+        If the passed transformer is an ADICVTransformer(), and the generated
+        series is a lumpy series; however, if there is no key matching "lumpy"
+        in the forecasters parameter, the fallback_forecaster will be used.
+        Additionally, if the fallback_forecaster is None, a ValueError will be thrown.
+        """
+        # passing time series through the provided transformer!
+
+        self.category_ = self.transformer_.fit_transform(X=y).iloc[0, 0]
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:224]: this should be explained in the docstring, i.e., that the first column is used.

Or, do we want to support multiple columns? In which case the keys could perhaps be tuples? You can get these via `itertuples`, and the length 1 case needs special treatment.

File: sktime/forecasting/compose/_category_compositor.py [hunk 19]
```diff
@@ -0,0 +1,479 @@
...
+            guaranteed to be of an mtype in self.get_tag("X_inner_mtype")
+            Exogeneous time series for the forecast
+        update_params : bool, optional (default=True)
+            whether model parameters should be updated
+
+        Returns
+        -------
+        self : reference to self
+        """
+        self.chosen_forecaster_.update(y=y, X=X, update_params=update_params)
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:316]: this should also be explained in the docstring, i.e., that the transformation is not updated.

or, we could make it a parameter whether the transformation updates, but perhaps not in this PR (seems fiddly)

File: sktime/forecasting/compose/_category_compositor.py [hunk 20]
```diff
@@ -0,0 +1,479 @@
...
+    Examples
+    --------
+    >>> from sktime.forecasting.compose._category_compositor import CategoryCompositor
+    >>> from sktime.forecasting.croston import Croston
+    >>> from sktime.forecasting.trend import PolynomialTrendForecaster
+    >>> from sktime.forecasting.naive import NaiveForecaster
+    >>> from sktime.transformations.series.adi_cv import ADICVTransformer
+    >>> from sktime.transformations.series.tests.test_adi_cv import (
+    ...     _generate_erratic_series)
+    >>> category_compositor = CategoryCompositor(
```
Comment (fkiraly) [inline sktime/forecasting/compose/_category_compositor.py:54]: nice example! would you like to add comments or explanation what this does? it is not obvious for the user, and a good example is accompanied by a good explanation (comments or text)

File: sktime/forecasting/compose/_category_compositor.py (hunk 12 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:18]: I like the sound of `ForecastByGroup`, I'll change the name!

File: sktime/forecasting/compose/_category_compositor.py (hunk 13 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:86]: Ah, got it. That makes sense. I'll ammend this and add it to the next commit.

File: sktime/forecasting/compose/_category_compositor.py (hunk 16 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:120]: I actually tried out `self._anytagis_then_set`, but I got the error that the function was not defined. Is it a defined function that the class should have inherited from `BaseForecaster`?

File: sktime/forecasting/compose/_category_compositor.py (hunk 20 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:54]: Ah, absolutely! I'll explain it in more detail. Sorry if this seems rushed, I've just been reading the LagLlama paper and have been very excited about getting started on adding that implementation as soon as I could... 😅

File: sktime/forecasting/compose/_category_compositor.py (hunk 14 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:94]: Got it, I'll have this change in the next commit.

File: sktime/forecasting/compose/_category_compositor.py (hunk 15 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:113]: Got it, I've set these to true for both `true_if_any_tags` and `true_if_all_tags`.

File: sktime/forecasting/compose/_category_compositor.py (hunk 18 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:224]: We could do it that way. However, would it then be a requirement that the first row must have the key that we are looking for? Also, if we are doing multiple columns (by which I assume you mean that the key could be in any column, but it would be in only in one specific column) would it be simpler if the user could pass an additional column name as a parameter where we should search for the key?

I'm okay with either design choice (user possibly specifies which column to choose or sets up the transformer such that the first column is the valid one), but could you tell me which approach might be easiest for the user?

File: sktime/forecasting/compose/_category_compositor.py (hunk 19 above)
Comment (shlok191) [reply sktime/forecasting/compose/_category_compositor.py:316]: I agree that it might be good to add that as a parameter (if the transformer updates, and be extension the chosen forecaster updates too). I'll explain this in the docstring for now but let me know if you'd like me to add this in for this PR!

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks, really great contribution!

I've reviewed the estimator, and the main thing I want to bring up: the forecasters for the components are not tunable via `get_params` etc. For that, we would have to use `_HeterogenousMetaEstimator`, which adds that feature - and also the "any tag" method that is missing.

The best template to use here would be `HierarchyEnsembleForecaster`, imo.

To use the ensemble base class, we may also have to change the `dict` syntax to list of `("name", forecaster, value)`, so the ensemble can be used.

Let me know what you think and whether this makes sense.

**Comment by shlok191** [comment]
@fkiraly, instead of using `HierarchyEnsembleForecaster` for which I am a little unsure the changes I might have to make to ensure that `_fit` works, could I possibly try to extend `_HeterogenousMetaEstimator`? I have read `HierarchyEnsembleForecaster`'s internal functions to assist with `get_params` and I think it might be easier for me to implement them for this forecaster!

**Comment by fkiraly** [comment]
> could I possibly try to extend `_HeterogenousMetaEstimator`

That's what I meant! Sorry if my phrasing was confusing as to the precise architecture.

I meant, the `HierarchyEnsembleForecaster` is a good example on how `_HeterogenousMetaEstimator` is meant to be used as an intermediate base class.

**Comment by shlok191** [comment]
Ah, got it! I thought I’d confirm. As a short update, I’ve completed the implementation locally and just have some bugs to fix to pass all the test cases. This should be done soon!

**Comment by fkiraly** [comment]
ready for review?

**Comment by shlok191** [comment]
@fkiraly, yes! I have tried to follow the lead from `HierarchyEnsembleForecaster` to implement the necessary functionality to ensure get_params and other related functions work as expected.

File: docs/source/api_reference/forecasting.rst [hunk 21]
```diff
@@ -50,6 +50,7 @@ Pipelines can also be constructed using ``*``, ``+``, and ``|`` dunders.
     ForecastByLevel
     Permute
     HierarchyEnsembleForecaster
+    CategoryCompositor
```
Comment (fkiraly) [inline docs/source/api_reference/forecasting.rst:53]: old name still here

File: sktime/forecasting/compose/_by_group_forecaster.py [hunk 22]
```diff
@@ -0,0 +1,526 @@
...
+from sktime.base._meta import _HeterogenousMetaEstimator
+from sktime.forecasting.base import BaseForecaster
+from sktime.forecasting.croston import Croston
+from sktime.forecasting.naive import NaiveForecaster
+from sktime.forecasting.trend import PolynomialTrendForecaster
+from sktime.transformations.base import BaseTransformer
+from sktime.transformations.series.adi_cv import ADICVTransformer
+
+
+class ByGroupForecaster(BaseForecaster, _HeterogenousMetaEstimator):
```
Comment (fkiraly) [inline sktime/forecasting/compose/_by_group_forecaster.py:15]: the name still feels a bit uninformtaive to me.

It should imply that the groups are computed from the time series. Not sure how to do that, that might be a good question to brainstorm.

File: sktime/forecasting/compose/_by_group_forecaster.py [hunk 23]
```diff
@@ -0,0 +1,526 @@
...
+            if category != "fallback_forecaster":
+                self.forecasters[category] = forecaster
+
+            else:
+                self.fallback_forecaster = forecaster
+
+
+# Function implementations that will be added dynamically
+# if the conditions are met. explained further above!
+def _update_predict_single(self, y, fh, X=None, update_params=True):
```
Comment (fkiraly) [inline sktime/forecasting/compose/_by_group_forecaster.py:408]: is this redundant? Why is this needed?

File: sktime/forecasting/compose/_by_group_forecaster.py [hunk 24]
```diff
@@ -0,0 +1,526 @@
...
+        where the category for which the respective forecaster is chosen
+        and the forecaster itself as the values for each tuple.
+
+        Returns
+        -------
+        forecasters : list[tuple[str, strsktime forecasters]]
+            The list of forecasters which is returned. Also includes the
+            fallback forecaster with the category: "fallback_forecaster"
+        """
+        return list(self.forecasters.values()) + [
```
Comment (fkiraly) [inline sktime/forecasting/compose/_by_group_forecaster.py:382]: should this be `items` rather than `values`?

File: sktime/forecasting/compose/_by_group_forecaster.py (hunk 23 above)
Comment (shlok191) [reply sktime/forecasting/compose/_by_group_forecaster.py:408]: Right, there isn't exactly much of an efficient improvement there. I'll remove this

File: sktime/forecasting/compose/_by_group_forecaster.py (hunk 22 above)
Comment (shlok191) [reply sktime/forecasting/compose/_by_group_forecaster.py:15]: I agree. I wanted to have a name that captures the 2-stage nature of this with something that could showcase that there is a `series-to-primitives` transformer and then a forecaster but I just cannot think of a name that is not a mouthful.

Should I drop this question in dev-chat? Maybe someone has some good ideas

File: sktime/forecasting/compose/_by_group_forecaster.py (hunk 24 above)
Comment (shlok191) [reply sktime/forecasting/compose/_by_group_forecaster.py:382]: Oh, that's right! I'll make the change

File: docs/source/api_reference/forecasting.rst (hunk 21 above)
Comment (shlok191) [reply docs/source/api_reference/forecasting.rst:53]: Sorry, I missed that, removing

File: sktime/forecasting/compose/_by_group_forecaster.py (hunk 22 above)
Comment (shlok191) [reply sktime/forecasting/compose/_by_group_forecaster.py:15]: How about something like `TransformSelectForecaster` or `PrimitivesDerivedForecaster`?

File: sktime/forecasting/compose/_by_group_forecaster.py (hunk 24 above)
Comment (fkiraly) [reply sktime/forecasting/compose/_by_group_forecaster.py:382]: I find it odd that the tests did not detect this, how about adding an extra test that checks that the `get_params` keys are as expected?

Further, I think you may want to check whether you have to coerce the keys to `str`, or whether it works without giving trouble (because this happens somewhere else already).

File: sktime/forecasting/compose/_by_group_forecaster.py (hunk 22 above)
Comment (shlok191) [reply sktime/forecasting/compose/_by_group_forecaster.py:15]: Sounds good! I'm planning on pushing in a new commit with the added changes soon and with the name `TransformSelectorForecaster` and I'll request folks to pitch in ideas on Discord. If we get something new I'll add it in!

Also, yes. I'll add an example too when I write the message!

File: sktime/forecasting/compose/_by_group_forecaster.py (hunk 24 above)
Comment (shlok191) [reply sktime/forecasting/compose/_by_group_forecaster.py:382]: Yes, I'm adding in that check with the next commit. Could you explain what you mean by coercing the keys to string? Is that in reference to the `"fallback_forecaster", self.fallback_forecaster` line where I specify fallback forecaster as a string key?

Comment (fkiraly) [reply sktime/forecasting/compose/_by_group_forecaster.py:382]: If you run `get_params` with `deep=True`, nested parameter keys will be strings `component__parametername`.

For the `_HeterogeneousMetaEstimator` to work, the name in name/estimator tuples needs to be string.

In cases where the transformer outputs non-string values (e.g., integers), in the property you may have to coerce the integers to strings - or maybe it is handled already somewhere.

File: sktime/forecasting/compose/_category_compositor.py (hunk 18 above)
Comment (fkiraly) [reply sktime/forecasting/compose/_category_compositor.py]: I would say, let's use the first column for now and worry about more columns later.

**Comment by shlok191** [comment]
@fkiraly, I have added in the changes we discussed! I have added tests to ensure that the parameters return from `get_params()` are as we hope they should be and that the `name` in the `(name, object)` tuples is are always strings.

Additionally, I have renamed the forecaster to `TransformSelectForecaster` and I am dropping a question in dev-chat about the update!

**Review by fkiraly — DISMISSED** [review]

File: sktime/forecasting/compose/_transform_select_forecaster.py [hunk 25]
```diff
@@ -0,0 +1,515 @@
...
+        Example:
+
+        If the passed transformer is an ADICVTransformer(), and the generated
+        series is a lumpy series; however, if there is no key matching "lumpy"
+        in the forecasters parameter, the fallback_forecaster will be used.
+        Additionally, if the fallback_forecaster is None, a ValueError will be thrown.
+        """
+        # passing time series through the provided transformer!
+
+        self.category_ = self.transformer_.fit_transform(X=y).iloc[0, 0]
```
Comment (fkiraly) [inline sktime/forecasting/compose/_transform_select_forecaster.py:233]: I think the args should be `(X=y, y=X)`, in case the transformer takes exogeneous data?

Comment (fkiraly) [reply sktime/forecasting/compose/_transform_select_forecaster.py]: added that, hope it does not break anything

Comment (shlok191) [reply sktime/forecasting/compose/_transform_select_forecaster.py]: Thank you so much! I just got off work and was just about to add it in. I'll keep an eye on the tests :)
